### DISCUSSION TITLE
ECJ not seeing generated classes/packages but javac does

### DISCUSSION THREAD

**Discussion by guw** [eclipse-jdt/eclipse.jdt.core#2076] [Q&A] [answered]
I need help with debugging a bit of a special problem with ECJ. I am using the ECJ compiler programmatically. 

I have two modes working:

1. via tool api 
https://github.com/salesforce/bazel-jdt-java-toolchain/blob/7e175dca3becc03ac4d9574382e56bee827c050f/compiler/src/main/buildjar/com/google/devtools/build/buildjar/javac/BlazeEcjToolMain.java#L99-L115

2. via extending `org.eclipse.jdt.internal.compiler.batch.Main` directly
https://github.com/salesforce/bazel-jdt-java-toolchain/blob/7e175dca3becc03ac4d9574382e56bee827c050f/compiler/src/main/buildjar/com/google/devtools/build/buildjar/javac/BlazeEcjMain.java#L321-L332

The second options allows me to implement some additional features (like tracking used dependencies ore generated sources).

I am running into a weird issue now, which works in the `javac` but does not work with ECJ in either mode. I wonder if someone has an idea what/where to debug.

Basically I have an annotation processor that is generating some Java code. Then there is also existing `.java` source files to be compiled in the same run, which already reference the "to be generated" code.

It works in `javac` but not with the tool interface or when extending Main the compiler. It looks like ECJ is unable to resolve the generated classes and reports a compile error (`The import foo.bar cannot be resolved`).

**Comment by guw** [comment, marked as answer]
To close the loop, I found the problem:
https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1654, fixed via https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1447

After building [a reproducible test case](https://github.com/salesforce/bazel-jdt-java-toolchain/commit/968fb8f219b34d7156cc98b5ec52b1d3492ed530) I was able to confirm it's fixed in 4.31RC1
